discojs/src/models/tfjs.ts

```diff
@@ -164,4 +214,123 @@ export class TFJS extends Model {
   extract (): tf.LayersModel {
     return this.model
   }
+
+  #batchToTF(batch: Batched<ModelEncoded[D]>): Record<"xs" | "ys", tf.Tensor> {
+    const outputSize = tf.util.sizeFromShape(
+      this.model.outputShape.map((dim) => {
+        if (Array.isArray(dim))
+          throw new Error("TODO support multiple outputs");
+        return dim ?? 1;
+      }),
+    );
+
+    switch (this.datatype) {
+      case "image": {
+        // cast as typescript doesn't reduce generic type
+        const b = batch as Batched<ModelEncoded["image"]>;
+
+        return tf.tidy(() => ({
+          xs: tf.stack(
+            b
+              .map(([image]) =>
+                tf.tensor3d(
+                  image.data,
+                  [image.width, image.height, 3],
+                  "float32",
+                ),
+              )
+              .toArray(),
+          ),
+          ys: tf.stack(
+            b
+              .map(([_, label]) => tf.oneHot(label, outputSize, 1, 0, "int32"))
+              .toArray(),
+          ),
+        }));
+      }
+      case "tabular": {
+        // cast as typescript doesn't reduce generic type
+        const b = batch as Batched<ModelEncoded["tabular"]>;
+
+        return tf.tidy(() => ({
+          xs: tf.stack(
+            b.map(([inputs, _]) => tf.tensor1d(inputs.toArray())).toArray(),
+          ),
+          ys: tf.stack(b.map(([_, output]) => tf.tensor1d([output])).toArray()),
+        }));
+      }
+      case "text": {
+        // cast as typescript doesn't reduce generic type
+        const b = batch as Batched<ModelEncoded["text"]>;
+
+        return {
+          xs: tf.stack(
+            b.map(([line]) => tf.tensor1d(line.toArray())).toArray(),
+          ),
+          ys: tf.stack(
+            b
+              .map(([line, next]) =>
+                tf.oneHot(line.shift().push(next).toArray(), outputSize),
+              )
+              .toArray(),
+          ),
+        };
```

Comment:
hoo indeed, indeed! dropping text support for TFJS.

btw, very comfortable change to implement, as types gets refined, one can simply change `class TFJS<D extends DataType>` to `class TFJS<D extends "image" | "tabular">` and works their way through the errors rather than letting it crash somewhere down the road.

Reply:
Haha nice reaping the benefits of your work!